### PR TITLE
head: provide series ID in ForEachSecondaryHash

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
related to https://github.com/grafana/mimir/issues/7976

### CODE REVIEW

File: tsdb/head.go [hunk 1]
```diff
@@ -2381,35 +2381,63 @@ func (h *Head) updateWALReplayStatusRead(current int) {
...
 
-		if len(buf) > 0 {
-			fn(buf)
+		if slices.len() > 0 {
+			fn(slices.slice1, slices.slice2)
 		}
 	}
 }
+
+type pairOfSlices[T1, T2 any] struct {
```
Comment (dimitarvdimitrov) [inline tsdb/head.go:2417]: all the operations on pairOfSlices are inlined so it shouldn't be different from just having two variables 


<details><summary>Details</summary>
<p>


```
tsdb/head.go:2379:31: (*Head).updateWALReplayStatusRead capturing by value: .autotmp_2 (addr=false assign=false width=8)
tsdb/head.go:2392:6: cannot inline (*Head).ForEachSecondaryHash: function too complex: cost 378 exceeds budget 80
tsdb/head.go:2392:7: parameter h leaks to {heap} with derefs=2:
tsdb/head.go:2392:7:   flow: sync.rw = **h:
tsdb/head.go:2392:7:     from h.series (dot of pointer) at tsdb/head.go:2409:4
tsdb/head.go:2392:7:     from h.series.locks (dot of pointer) at tsdb/head.go:2409:11
tsdb/head.go:2392:7:     from h.series.locks[i] (dot of pointer) at tsdb/head.go:2409:17
tsdb/head.go:2392:7:     from h.series.locks[i].RWMutex (dot) at tsdb/head.go:2409:20
tsdb/head.go:2392:7:     from h.series.locks[i].RWMutex (address-of) at tsdb/head.go:2409:20
tsdb/head.go:2392:7:     from sync.rw := h.series.locks[i].RWMutex (assign-pair) at tsdb/head.go:2409:28
tsdb/head.go:2392:7:   flow: {heap} = sync.rw:
tsdb/head.go:2392:7:     from (*sync.RWMutex).rUnlockSlow(sync.rw, sync.r) (call parameter) at tsdb/head.go:2409:28
tsdb/head.go:2392:7: leaking param content: h
tsdb/head.go:2392:37: fn does not escape
tsdb/head.go:2393:57: inlining call to newPairOfSlices[go.shape.uint64,go.shape.uint32]
tsdb/head.go:2393:57: make([]go.shape.uint64, length) escapes to heap:
tsdb/head.go:2393:57:   flow: {heap} = &{storage for make([]go.shape.uint64, length)}:
tsdb/head.go:2393:57:     from make([]go.shape.uint64, length) (non-constant size) at tsdb/head.go:2393:57
tsdb/head.go:2393:57: make([]go.shape.uint32, length) escapes to heap:
tsdb/head.go:2393:57:   flow: {heap} = &{storage for make([]go.shape.uint32, length)}:
tsdb/head.go:2393:57:     from make([]go.shape.uint32, length) (non-constant size) at tsdb/head.go:2393:57
tsdb/head.go:2393:57: make([]go.shape.uint64, length) escapes to heap
tsdb/head.go:2393:57: make([]go.shape.uint32, length) escapes to heap
tsdb/head.go:2396:24: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].reset
tsdb/head.go:2398:26: inlining call to sync.(*RWMutex).RLock
tsdb/head.go:2398:26: inlining call to atomic.(*Int32).Add
tsdb/head.go:2401:26: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].append
tsdb/head.go:2406:27: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].append
tsdb/head.go:2409:28: inlining call to sync.(*RWMutex).RUnlock
tsdb/head.go:2409:28: inlining call to atomic.(*Int32).Add
tsdb/head.go:2411:16: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].len
tsdb/head.go:2422:6: can inline newPairOfSlices[go.shape.uint64,go.shape.uint32] with cost 8 as: func(*[5]uintptr, int) pairOfSlices[go.shape.uint64,go.shape.uint32] { return pairOfSlices[go.shape.uint64,go.shape.uint32]{...} }
tsdb/head.go:2422:6: can inline newPairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] with cost 14 as: func(int) pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] { return newPairOfSlices[go.shape.uint64,go.shape.uint32](&.dict.newPairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32], length) }
tsdb/head.go:2422:6: inlining call to newPairOfSlices[go.shape.uint64,go.shape.uint32]
tsdb/head.go:2422:6: make([]go.shape.uint64, length) escapes to heap:
tsdb/head.go:2422:6:   flow: {heap} = &{storage for make([]go.shape.uint64, length)}:
tsdb/head.go:2422:6:     from make([]go.shape.uint64, length) (non-constant size) at tsdb/head.go:2422:6
tsdb/head.go:2422:6: make([]go.shape.uint32, length) escapes to heap:
tsdb/head.go:2422:6:   flow: {heap} = &{storage for make([]go.shape.uint32, length)}:
tsdb/head.go:2422:6:     from make([]go.shape.uint32, length) (non-constant size) at tsdb/head.go:2422:6
tsdb/head.go:2422:6: make([]go.shape.uint64, length) escapes to heap
tsdb/head.go:2422:6: make([]go.shape.uint32, length) escapes to heap
tsdb/head.go:2424:15: make([]go.shape.uint64, length) escapes to heap:
tsdb/head.go:2424:15:   flow: {heap} = &{storage for make([]go.shape.uint64, length)}:
tsdb/head.go:2424:15:     from make([]go.shape.uint64, length) (non-constant size) at tsdb/head.go:2424:15
tsdb/head.go:2424:15: make([]go.shape.uint64, length) escapes to heap
tsdb/head.go:2425:15: make([]go.shape.uint32, length) escapes to heap:
tsdb/head.go:2425:15:   flow: {heap} = &{storage for make([]go.shape.uint32, length)}:
tsdb/head.go:2425:15:     from make([]go.shape.uint32, length) (non-constant size) at tsdb/head.go:2425:15
tsdb/head.go:2425:15: make([]go.shape.uint32, length) escapes to heap
tsdb/head.go:2429:6: can inline pairOfSlices[go.shape.uint64,go.shape.uint32].reset with cost 16 as: method(pairOfSlices[go.shape.uint64,go.shape.uint32]) func(*[5]uintptr) pairOfSlices[go.shape.uint64,go.shape.uint32] { p.slice1 = p.slice1[:0]; p.slice2 = p.slice2[:0]; return p }
tsdb/head.go:2429:6: can inline pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32].reset with cost 21 as: method(pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32]) func() pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] { return pairOfSlices[go.shape.uint64,go.shape.uint32].reset(p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32]) }
tsdb/head.go:2429:6: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].reset
tsdb/head.go:2429:7: parameter p leaks to ~r0 with derefs=0:
tsdb/head.go:2429:7:   flow: ~r0 = p:
tsdb/head.go:2429:7:     from return p (return) at tsdb/head.go:2432:2
tsdb/head.go:2429:7: parameter p leaks to ~r0 with derefs=0:
tsdb/head.go:2429:7:   flow: p = p:
tsdb/head.go:2429:7:     from p, .dict := p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] (assign-pair) at tsdb/head.go:2429:6
tsdb/head.go:2429:7:   flow: ~R0 = p:
tsdb/head.go:2429:7:     from ~R0 = p (assign-pair) at tsdb/head.go:2429:6
tsdb/head.go:2429:7:   flow: ~r0 = ~R0:
tsdb/head.go:2429:7:     from return ~R0 (return) at tsdb/head.go:2429:6
tsdb/head.go:2429:7: parameter .this leaks to ~r0 with derefs=1:
tsdb/head.go:2429:7:   flow: p = *.this:
tsdb/head.go:2429:7:     from .this (indirection) at <autogenerated>:1
tsdb/head.go:2429:7:     from p := .this (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: p = p:
tsdb/head.go:2429:7:     from p, .dict := p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: ~R0 = p:
tsdb/head.go:2429:7:     from ~R0 = p (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: ~R0 = ~R0:
tsdb/head.go:2429:7:     from ~R0 = ~R0 (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: ~r0 = ~R0:
tsdb/head.go:2429:7:     from return ~R0 (return) at <autogenerated>:1
tsdb/head.go:2429:7: parameter .this leaks to ~r0 with derefs=1:
tsdb/head.go:2429:7:   flow: p = *.this:
tsdb/head.go:2429:7:     from .this (indirection) at <autogenerated>:1
tsdb/head.go:2429:7:     from p, .dict := .this, .dict (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: ~R0 = p:
tsdb/head.go:2429:7:     from ~R0 = p (assign-pair) at <autogenerated>:1
tsdb/head.go:2429:7:   flow: ~r0 = ~R0:
tsdb/head.go:2429:7:     from return ~R0 (return) at <autogenerated>:1
tsdb/head.go:2435:6: can inline pairOfSlices[go.shape.uint64,go.shape.uint32].append with cost 16 as: method(pairOfSlices[go.shape.uint64,go.shape.uint32]) func(*[5]uintptr, go.shape.uint64, go.shape.uint32) pairOfSlices[go.shape.uint64,go.shape.uint32] { p.slice1 = append(p.slice1, t1); p.slice2 = append(p.slice2, t2); return p }
tsdb/head.go:2435:6: can inline pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32].append with cost 23 as: method(pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32]) func(chunks.HeadSeriesRef, uint32) pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32] { return pairOfSlices[go.shape.uint64,go.shape.uint32].append(p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32], t1, t2) }
tsdb/head.go:2435:6: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].append
tsdb/head.go:2435:7: parameter p leaks to ~r0 with derefs=0:
tsdb/head.go:2435:7:   flow: ~r0 = p:
tsdb/head.go:2435:7:     from return p (return) at tsdb/head.go:2438:2
tsdb/head.go:2435:7: parameter p leaks to ~r0 with derefs=0:
tsdb/head.go:2435:7:   flow: p = p:
tsdb/head.go:2435:7:     from p, .dict, t1, t2 := p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32], t1, t2 (assign-pair) at tsdb/head.go:2435:6
tsdb/head.go:2435:7:   flow: ~R0 = p:
tsdb/head.go:2435:7:     from ~R0 = p (assign-pair) at tsdb/head.go:2435:6
tsdb/head.go:2435:7:   flow: ~r0 = ~R0:
tsdb/head.go:2435:7:     from return ~R0 (return) at tsdb/head.go:2435:6
tsdb/head.go:2435:7: parameter .this leaks to ~r0 with derefs=1:
tsdb/head.go:2435:7:   flow: p = *.this:
tsdb/head.go:2435:7:     from .this (indirection) at <autogenerated>:1
tsdb/head.go:2435:7:     from p, t1, t2 := .this, t1, t2 (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: p = p:
tsdb/head.go:2435:7:     from p, .dict, t1, t2 := p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32], t1, t2 (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: ~R0 = p:
tsdb/head.go:2435:7:     from ~R0 = p (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: ~R0 = ~R0:
tsdb/head.go:2435:7:     from ~R0 = ~R0 (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: ~r0 = ~R0:
tsdb/head.go:2435:7:     from return ~R0 (return) at <autogenerated>:1
tsdb/head.go:2435:7: parameter .this leaks to ~r0 with derefs=1:
tsdb/head.go:2435:7:   flow: p = *.this:
tsdb/head.go:2435:7:     from .this (indirection) at <autogenerated>:1
tsdb/head.go:2435:7:     from p, .dict, t1, t2 := .this, .dict, t1, t2 (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: ~R0 = p:
tsdb/head.go:2435:7:     from ~R0 = p (assign-pair) at <autogenerated>:1
tsdb/head.go:2435:7:   flow: ~r0 = ~R0:
tsdb/head.go:2435:7:     from return ~R0 (return) at <autogenerated>:1
tsdb/head.go:2441:6: can inline pairOfSlices[go.shape.uint64,go.shape.uint32].len with cost 4 as: method(pairOfSlices[go.shape.uint64,go.shape.uint32]) func(*[5]uintptr) int { return len(p.slice1) }
tsdb/head.go:2441:6: can inline pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32].len with cost 9 as: method(pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32]) func() int { return pairOfSlices[go.shape.uint64,go.shape.uint32].len(p, &.dict.pairOfSlices[github.com/prometheus/prometheus/tsdb/chunks.HeadSeriesRef,uint32]) }
tsdb/head.go:2441:6: inlining call to pairOfSlices[go.shape.uint64,go.shape.uint32].len
```

</p>
</details> 

Also, nothing new escapes to the heap


<details><summary>Details</summary>
<p>


```
github.com/prometheus/prometheus/tsdb.(*Head).ForEachSecondaryHash
/Users/dimitar/grafana/forks/mimir-prometheus/tsdb/head.go

  Total:     24.58MB    24.58MB (flat, cum)   0.4%
   2388            .          .           // 
   2389            .          .           // No locks are held when function is called. 
   2390            .          .           // 
   2391            .          .           // Slices passed to the function are reused between calls. 
   2392            .          .           func (h *Head) ForEachSecondaryHash(fn func(ref []chunks.HeadSeriesRef, secondaryHash []uint32)) { 
   2393      24.58MB    24.58MB           	slices := newPairOfSlices[chunks.HeadSeriesRef, uint32](512) 
   2394            .          .            
   2395            .          .           	for i := 0; i < h.series.size; i++ { 
   2396            .          .           		slices = slices.reset() 
   2397            .          .            
   2398            .          .           		h.series.locks[i].RLock() 
github.com/prometheus/prometheus/tsdb.newPairOfSlices[go.shape.uint64,go.shape.uint32]
/Users/dimitar/grafana/forks/mimir-prometheus/tsdb/head.go

  Total:     24.58MB    24.58MB (flat, cum)   0.4%
   2419            .          .           	slice2 []T2 
   2420            .          .           } 
   2421            .          .            
   2422            .          .           func newPairOfSlices[T1, T2 any](length int) pairOfSlices[T1, T2] { 
   2423            .          .           	return pairOfSlices[T1, T2]{ 
   2424      18.07MB    18.07MB           		slice1: make([]T1, length), 
   2425       6.51MB     6.51MB           		slice2: make([]T2, length), 
   2426            .          .           	} 
   2427            .          .           } 
   2428            .          .            
   2429            .          .           func (p pairOfSlices[T1, T2]) reset() pairOfSlices[T1, T2] { 
   2430            .          .           	p.slice1 = p.slice1[:0] 
```

</p>
</details>

File: tsdb/head.go [hunk 2]
```diff
@@ -2381,35 +2381,63 @@ func (h *Head) updateWALReplayStatusRead(current int) {
 	h.stats.WALReplayStatus.Current = current
 }
 
-// ForEachSecondaryHash iterates over all series in the Head, and passes secondary hashes of the series
-// to the function. Function is called with batch of hashes, in no specific order. Hash for each series
-// in the head is included exactly once. Series for corresponding hash may be deleted while the function
-// is running, and series inserted while this function runs may be reported or ignored.
+// ForEachSecondaryHash iterates over all series in the Head, and passes references and secondary hashes of the series
+// to the function. Function is called with batch of refs and hashes, in no specific order. The order of the refs
+// in the same as the order of the hashes. Each seriesin the head is included exactly once.
```
Comment (pstibrany) [inline tsdb/head.go:2386]:
```suggestion
// in the same as the order of the hashes. Each series in the head is included exactly once.
```

File: tsdb/head.go (hunk 1 above)
Comment (pstibrany) [reply tsdb/head.go:2417]: introducing new type adds lot more boilerplate lines than just using two slices in the first place :)

**Review by pstibrany — APPROVED** [review]

File: tsdb/head.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply tsdb/head.go:2417]: i don't know.. that way the function keeps growing and growing and following it becomes harder and harder. It also makes it easier to make a mistake in the future - just appending to one of the slices. If that doesn't bother you I can remove the pairOfSlices 😄

Comment (pstibrany) [reply tsdb/head.go:2417]: I would expect that test would catch such mistakes. That said, feel free to use this version, it’s not a blocker for me.

**Pull request merged**